Tutorial 4 - Geophysics (Seismology): Add JN

yvonnefroehlich opened the pull request:
This PR adds the Jupyter notebook for **Tutorial 4 - Geophysics (Seismology)**.

- **Study region**: Chilean subduction zone since there are significant topographic variation, earthquakes, and volcanoes
- **Orientation**: https://www.pygmt.org/dev/gallery/images/cross_section.html

Content:
- [x] Prepare **gridded data**: [`pygmt.xyz2grd`](https://www.pygmt.org/v0.13.0/api/generated/pygmt.grd2xyz.html)
- [x] Create a **contour map**: [`pygmt.Figure.grdcontour`](https://www.pygmt.org/v0.13.0/api/generated/pygmt.Figure.grdcontour.html)
- [x] Create a **profile plot**: [`pygmt.project`](https://www.pygmt.org/v0.13.0/api/generated/pygmt.project.html) and [`pygmt.grdtrack`](https://www.pygmt.org/v0.13.0/api/generated/pygmt.grdtrack.html)
- [x] Additional features: seismicity, beachball, plate motion 


Feel free to fill the subsections with suggestions for code and docs🙂.

**Preview**:
- https://github.com/GenericMappingTools/agu24workshop/blob/add-jn-geo-seis/book/tut4_geophysics.ipynb
- https://gitnotebooks.com/GenericMappingTools/agu24workshop/pull/8

Fixes: #6

yvonnefroehlich committed:
Add JN for geophysics with basic structure

yvonnefroehlich committed:
Make headings consistent

yvonnefroehlich committed:
Improve introduction cell | Add section 'Addional comments'

yvonnefroehlich committed:
Improve structure

jhtong33 commented:
Hi @yvonnefroehlich 
do we use `grdproject` or `project` ? 
Since `grdproject` is specified for a region rather than individual point-to-point. 

see also: https://www.pygmt.org/v0.13.0/gallery/images/cross_section.html#sphx-glr-gallery-images-cross-section-py

Commit:
code for grdcontour, project, and grdtrack, need to add note

yvonnefroehlich committed:
Update '_toc.yml'

Commit:
lack of xyz2grd, and need to add last part note

yvonnefroehlich commented:
> Hi @yvonnefroehlich do we use `grdproject` or `project` ? Since `grdproject` is specified for a region rather than individual point-to-point.
> 
> see also: https://www.pygmt.org/v0.13.0/gallery/images/cross_section.html#sphx-glr-gallery-images-cross-section-py

Yep, I am sorry, I mean `project` I mistyped this 🙁.

yvonnefroehlich committed:
Fix typo - 'grdproject' to 'project'

yvonnefroehlich committed:
Improve coding style

yvonnefroehlich committed:
Fix typo

yvonnefroehlich committed:
Improve code

yvonnefroehlich committed:
Reduce image resolution

yvonnefroehlich committed:
Use list

yvonnefroehlich reviewed (commented):
Thanks a lot, @jhtong33 , for getting started here 👍!

I made a few updates:
- I prefer to use the Scientific colour maps by F. Crameri, i.e., for topography, "oleron".
- I reduced the resolution of the image showing within the Jupyter notebook to keep the file size smaller.
- I figured out that list input is supported for the `center` and `endpoint` parameters of `project`.
- Fix typo `grdproject` to `project`.
- So minor code style improvements.
- I think we should be consistent and use either single (`'`) or double (`"`) quotation marks; I prefer double quotation marks.

Commit:
pygmt.xyz2grd finish

jhtong33 commented:
> Thanks a lot, @jhtong33 , for getting started here 👍!
> 
> I made a few updates:
> 
> * I prefer to use the Scientific colour maps by F. Crameri, i.e., for topography, "oleron".
> * I reduced the resolution of the image showing within the Jupyter notebook to keep the file size smaller.
> * I figured out that list input is supported for the `center` and `endpoint` parameters of `project`.
> * Fix typo `grdproject` to `project`.
> * So minor code style improvements.
> * I think we should be consistent and use either single (`'`) or double (`"`) quotation marks; I prefer double quotation marks.

Great! Thanks for your updates. 
I used a remote dataset to convert data into a .txt file, which I then provided as input materials for `pygmt.xyz2grd`.

yvonnefroehlich committed:
Make argument for yshift flexible

yvonnefroehlich committed:
Add section for adding beachball

yvonnefroehlich committed:
Move beachball part at the end of the script

yvonnefroehlich committed:
Update comments | Improve coding style

Commit:
material for seismicity

Commit:
add seismicity code, need to add note

jhtong33 commented:
Hi @yvonnefroehlich 
I have one idea that plotting beachballs for magnitude >6 events to make the tutorial more comprehensive. 
However, If time make it challenging to cover all details within a 45-minute session, it would be great to focus on one significant event as an example.

see also: https://www.isc.ac.uk/cgi-bin/web-db-run?request=COMPREHENSIVE&out_format=CATCSV&searchshape=RECT&bot_lat=-40&top_lat=0&left_lon=-90&right_lon=-60&ctr_lat=&ctr_lon=&radius=&max_dist_units=deg&srn=&grn=&start_year=2022&start_month=1&start_day=01&start_time=00%3A00%3A00&end_year=2023&end_month=1&end_day=01&end_time=00%3A00%3A00&min_dep=&max_dep=&min_mag=6&max_mag=&req_mag_type=MW&req_mag_agcy=&include_links=on

yvonnefroehlich committed:
Improve coding style | Use SC by Crameri

yvonnefroehlich committed:
Fix typos

yvonnefroehlich committed:
Improve and add docs

yvonnefroehlich committed:
Add links

yvonnefroehlich commented:
> Hi @yvonnefroehlich I have one idea that plotting beachballs for magnitude >6 events to make the tutorial more comprehensive. However, If time make it challenging to cover all details within a 45-minute session, it would be great to focus on one significant event as an example.
> 
> see also: https://www.isc.ac.uk/cgi-bin/web-db-run?request=COMPREHENSIVE&out_format=CATCSV&searchshape=RECT&bot_lat=-40&top_lat=0&left_lon=-90&right_lon=-60&ctr_lat=&ctr_lon=&radius=&max_dist_units=deg&srn=&grn=&start_year=2022&start_month=1&start_day=01&start_time=00%3A00%3A00&end_year=2023&end_month=1&end_day=01&end_time=00%3A00%3A00&min_dep=&max_dep=&min_mag=6&max_mag=&req_mag_type=MW&req_mag_agcy=&include_links=on

Regarding plotting the seismicity as color- and size-coded circles: Actually, is more or less the same as I do already in part "1.3 Create a geographical map of the earthquakes" of tutorial 2 (please see https://github.com/GenericMappingTools/agu24workshop/blob/add-jn-pd-gpd/book/tut2_spec_pd_gpd.ipynb). Reading data in to a pandas.Dataframe and plotting it with color- and size-coding. Maybe we just provide the data and they should try to write the code on their own as an additional task?

Some updates from my side
 - Code for plotting one single earthquake as beachball
 - More documentation
 - Links to the PyGMT documentation for the used functions and methods
 - Usage of SC by F. Crameri
 - Reduced resolution for all images shown within the JN to keep the file small

jhtong33 commented:
> Regarding plotting the seismicity as color- and size-coded circles: Actually, is more or less the same as I do already in part "1.3 Create a geographical map of the earthquakes" of tutorial 2 (please see https://github.com/GenericMappingTools/agu24workshop/blob/add-jn-pd-gpd/book/tut2_spec_pd_gpd.ipynb). Reading data in to a pandas.Dataframe and plotting it with color- and size-coding. Maybe we just provide the data and they should try to write the code on their own as an additional task?

Oh! Great! I agree that they can write the code on their own. Perhaps we can write more details or provide the example figure with them. How do you think ?

yvonnefroehlich committed:
Fix typos

Commit:
fix typo

Commit:
typo

Commit:
typo

yvonnefroehlich committed:
Use colormap with was not flipped between SC versions | TEST Comment 'fig.show'

yvonnefroehlich commented on book/tut4_geophysics.ipynb:
```diff
@@ -0,0 +1,1388 @@
+{
```
TEST: Please change "gray" to "lightgray"

yvonnefroehlich replied:
Updated in commit https://github.com/GenericMappingTools/agu24workshop/pull/8/commits/04de4e7ca14b43555a9d9beaa677dfbd9e271ac6.

yvonnefroehlich commented on book/tut4_geophysics.ipynb:
```diff
@@ -0,0 +1,1388 @@
+{
```
_Commented on notebook `book/tut4_geophysics.ipynb` Cell 37 Line 1_
```python
pygmt.makecpt(cmap="acton", series=[0, 500], reverse=True)
```
TEST: Use another colormap, as the SC colormap "lajolla" flips between the different (GMT) versions.

yvonnefroehlich replied:
Changed in commit https://github.com/GenericMappingTools/agu24workshop/pull/8/commits/334ef23b3791a366f59de3ee698034344dbe58f2.

yvonnefroehlich committed:
TEST Add 'Figure.show' back

yvonnefroehlich committed:
Fix typo in code

yvonnefroehlich committed:
Introduce variable for resoultion of displayed images

yvonnefroehlich committed:
Use correct color in comment

yvonnefroehlich committed:
Add comment for low resolution of images within JN

yvonnefroehlich committed:
Test resolution of 100 dpi

yvonnefroehlich committed:
Use resolution of 150 dpi for images displayed in JN

yvonnefroehlich committed:
Show only final images within JN

yvonnefroehlich commented on book/tut4_geophysics.ipynb:
```diff
@@ -0,0 +1,1397 @@
+{
```
TEST: Maybe introduce a variable at the beginning to have always the same value and be able to change it for all occurrences identically.

yvonnefroehlich replied:
Done in commit https://github.com/GenericMappingTools/agu24workshop/pull/8/commits/477406eb81203b3e50715f895cdfb9b7c9232349.

yvonnefroehlich committed:
Improve 'Plotting addinitional feature' section

yvonnefroehlich committed:
Add section 'Orientation / suggestion for 4.1'

yvonnefroehlich committed:
Improve docs of 'Addiional feature section'

yvonnefroehlich committed:
Add links to PyGMT

yvonnefroehlich committed:
Fix copy past error

yvonnefroehlich commented:
In this tutorial, we create a more complex figure across multiple cells. In this case, re-running cells to play around with different arguments for the parameters does not work (it only works for the cell that contains the creation of the new Figure instance).

To improve this, I was wondering if it can work to create for each cell a new "figure variable",  i.e. start with `fig1 = pygmt.Figure()`, next cell `fig2 = fig1`, next cell `fig3 = fig2`, etc. (please see commit https://github.com/GenericMappingTools/agu24workshop/pull/8/commits/075ea2e5f63b01ab24ecf0b06351def5ca166eac for this change). In this way, `fig1` is not affected by the changes in the cell for `fig2` and re-running this cell would again start at the status of `fig1`. But so far, it looks like it does not work as expected, and both `fig1` and `fig2` are changed. I am not a Jupyter notebook or lab user (but prefer to write normal [Python] scripts), so ideas from more experienced users are highly welcome 🙂.

yvonnefroehlich committed:
Add variables 'fig*'

yvonnefroehlich committed:
Add color-coding for profil | Update and Improve comments

yvonnefroehlich committed:
Flip order of '4.1seismicity' and '4.2 beachball'

yvonnefroehlich committed:
Add subsections for section 3 to improve structur

yvonnefroehlich committed:
Reduce symbole size

yvonnefroehlich commented:
So, some more updates

- **Section 3**: Added subsection for Section 3 to improve the structure
- **Section 4**: Removed the code for plotting seismicity and added some references for orientation
- Some improvements and additions to the docs
- Use a colormap of the SC that does not flip between the (GMT) versions
- Change color of contour line labels for better visibility
- Introduces a variable to set the resolution for the images within the JN for all images consistently
- **Cartesian plot**: Plot the profile with color-coding for the elevation

yvonnefroehlich committed:
Fix typo

yvonnefroehlich committed:
Fix typos

Commit:
11/24

jhtong33 commented:
> In this way, `fig1` is not affected by the changes in the cell for `fig2` and re-running this cell would again start at the status of `fig1`. But so far, it looks like it does not work as expected, but both `fig1` and `fig2` are changed. I am not a Jupyter notebook or lab user (but prefer to write normal [Python] scripts), so ideas from more experienced users are highly welcome 🙂.

Understanding your concerns, this approach achieves the same results we expected. By doing it this way, we can avoid altering the original figure.

> * **Cartesian plot**: Plot the profile with color-coding for the elevation

I prefer using a black line since color-coding does not better visualize the topography variations. The y-axis already effectively indicates this information.

Commit:
change to black line, minor typos

jhtong33 commented:
Hi @yvonnefroehlich, I think this tutorial can be reviewed.

yvonnefroehlich committed:
Fix typo

yvonnefroehlich committed:
Change back to 1 point thick line and use red color to fit to profile

yvonnefroehlich commented:
> > * **Cartesian plot**: Plot the profile with color-coding for the elevation
> 
> I prefer using a black line since color-coding does not better visualize the topography variations. The y-axis already effectively indicates this information.

Totally fine! Maybe using the same color as the profil line has would nice (in our case red).

yvonnefroehlich commented:
> > In this way, `fig1` is not affected by the changes in the cell for `fig2` and re-running this cell would again start at the status of `fig1`. But so far, it looks like it does not work as expected, but both `fig1` and `fig2` are changed. I am not a Jupyter notebook or lab user (but prefer to write normal [Python] scripts), so ideas from more experienced users are highly welcome 🙂.
> 
> Understanding your concerns, this approach achieves the same results we expected. By doing it this way, we can avoid altering the original figure.

Hm. But for me, this approach does not work in the expected way. Actually, this reminds me a bit of ObsPy where users actively need to copy the Stream object, before applying any processing steps in case they want to keep their raw data in the initial  variable unchanged (https://docs.obspy.org/packages/autogen/obspy.core.stream.Stream.copy.html).

yvonnefroehlich committed:
Add path to SCM

yvonnefroehlich committed:
Make lables in profil plot consistent to map

yvonnefroehlich committed:
Use name instead of number for font

yvonnefroehlich committed:
Re-run JN

yvonnefroehlich committed:
Remove empty cell at the end of the JN

jhtong33 committed:
tong tmp

jhtong33 commented:
> Hm. But for me, this approach does not work in the expected way. Actually, this reminds me a bit of ObsPy where users actively need to copy the Stream object, before applying any processing steps in case they want to keep their raw data in the initial variable unchanged (https://docs.obspy.org/packages/autogen/obspy.core.stream.Stream.copy.html).

I misunderstood your meaning previously. I also encounter same condition in Jupyter lab. 
In obspy, they use copy.deepcopy(self). I tried to use same way to solve this issue, but the result is same.

yvonnefroehlich committed:
Remove numbers for Figure as not working

yvonnefroehlich committed:
Write 'Note' consistently the same

yvonnefroehlich commented:
> > Hm. But for me, this approach does not work in the expected way. Actually, this reminds me a bit of ObsPy where users actively need to copy the Stream object, before applying any processing steps in case they want to keep their raw data in the initial variable unchanged (https://docs.obspy.org/packages/autogen/obspy.core.stream.Stream.copy.html).
> 
> I misunderstood your meaning previously. I also encounter same condition in Jupyter lab. In obspy, they use copy.deepcopy(self). I tried to use same way to solve this issue, but the result is same.

No worries!
As this is not working as expected, I removed introducing different variables for the Figure in https://github.com/GenericMappingTools/agu24workshop/pull/8/commits/2a2e2382855617be1224e625ee89521585efa07a.

yvonnefroehlich committed:
Rename JN - add leading 0

yvonnefroehlich committed:
Fix typo

yvonnefroehlich committed:
Merge branch 'main' into add-jn-geo-seis

yvonnefroehlich committed:
Merge branch 'main' into add-jn-geo-seis

yvonnefroehlich commented:
> > Hm. But for me, this approach does not work in the expected way. Actually, this reminds me a bit of ObsPy where users actively need to copy the Stream object, before applying any processing steps in case they want to keep their raw data in the initial variable unchanged (https://docs.obspy.org/packages/autogen/obspy.core.stream.Stream.copy.html).
> 
> I misunderstood your meaning previously. I also encounter same condition in Jupyter lab. In obspy, they use copy.deepcopy(self). I tried to use same way to solve this issue, but the result is same.

This seem to be more complicated to implemente, please see https://github.com/GenericMappingTools/pygmt/issues/3645#issuecomment-2499892184 for details.

yvonnefroehlich committed:
Rename 'spec' to 'spe'

yvonnefroehlich commented:
Are there any major review requests on this tutorial 🙂? Otherwise, I feel we can move on and merge this tutorial in the current state and, if needed, open separat PRs for minor corrections and improvements.

Esteban82 commented:
I have two observations only: 
1. The first row of the `seismicity_2022.csv` file is: `time,lon,lat,depth,mag`. With GMT, it is good to add a `#` to the beginning of the line so that it is automatically identified as a header. I suspect that with PyGMT this is not necessary. is it true?
2. It would be possible to add a shading effect to the topographic map of South America. The idea would be to show that with GMT you can make very good maps simply.
Then, during the workshop it might be better to say that one removes the shading effect to make the processing faster.

yvonnefroehlich commented:
> 1. The first row of the `seismicity_2022.csv` file is: `time,lon,lat,depth,mag`. With GMT, it is good to add a `#` to the beginning of the line so that it is automatically identified as a header. I suspect that with PyGMT this is not necessary. is it true?

In this tutorial, we read the CSV file via pandas into a DataFrame. The first line is by default considered as the header line and used for the column names. Later, these column names can be used to access the single columns. pandas does not require a `#` to indicate a header line,  but would include the `#` into the first column name 😬:

```                            
                            #time      lon      lat    depth  mag
0     2022-12-31T23:04:12.784000Z -69.6800 -16.5378  196.743  4.9
1     2022-12-31T12:47:57.842000Z -71.5625 -32.2686   48.300  4.5
2     2022-12-31T03:31:43.830000Z -68.8969 -23.1036   96.224  5.1
3     2022-12-31T00:43:57.999000Z -77.7617  -1.7451  182.543  4.3
4     2022-12-30T22:35:25.866000Z -67.1397 -24.0267  232.941  4.5
                          ...      ...      ...      ...  ...
```
Thus, I prefer to leave this file as is.

yvonnefroehlich commented:
> 2. It would be possible to add a shading effect to the topographic map of South America. The idea would be to show that with GMT you can make very good maps simply.
>    Then, during the workshop it might be better to say that one removes the shading effect to make the processing faster.

Hm. You mean something simliar as Max does in the Xarray tutorial in the section "Processing raster data" (https://github.com/GenericMappingTools/agu24workshop/blob/add-xarray-tutorial/book/tut03_spec_xarray.ipynb) using `grdgradient`? Maybe we can include this in the section "Additional comments" as the tutorial is already a bit longer?

yvonnefroehlich committed:
Clear cells

Esteban82 commented:
> Hm. You mean something simliar as Max does in the Xarray tutorial in the section "Processing raster data" (https://github.com/GenericMappingTools/agu24workshop/blob/add-xarray-tutorial/book/tut03_spec_xarray.ipynb) using `grdgradient`? Maybe we can include this in the section "Additional comments" as the tutorial is already a bit longer?

My idea is just to make the maps look better. I don't think it's necessary to explain anything. So, I think you can add something like `shading="+a"` in the code so the map looks better.

yvonnefroehlich committed:
Merge branch 'main' into add-jn-geo-seis

yvonnefroehlich committed:
Add code and docs for 'shading' parameter

yvonnefroehlich commented:
> > Hm. You mean something simliar as Max does in the Xarray tutorial in the section "Processing raster data" (https://github.com/GenericMappingTools/agu24workshop/blob/add-xarray-tutorial/book/tut03_spec_xarray.ipynb) using `grdgradient`? Maybe we can include this in the section "Additional comments" as the tutorial is already a bit longer?
> 
> My idea is just to make the maps look better. I don't think it's necessary to explain anything. So, I think you can add something like `shading="+a"` in the code so the map looks better.

Ah, I first thought about `grdgradient` but applying the illumination directly via `grdimage` sounds good to me. Made an update in commit https://github.com/GenericMappingTools/agu24workshop/pull/8/commits/a3ded9d26f7c2b9c853221cbb8a05fc36425eaac.

yvonnefroehlich commented on book/tut04_geophysics.ipynb:
```diff
@@ -0,0 +1,694 @@
+    "3. `annotation`: Annotate contour levels.\n",
+    "4. `limit`: Draw contours below low or above high, e.g., [-4000, 0] means drawing contour lines below sea level and above -4000 m.\n",
+    "\n",
+    "Before we plot the grid as contour lines, we plot the grid with color-coding using [`pygmt.Figure.grdimage`](https://www.pygmt.org/v0.13.0/api/generated/pygmt.Figure.grdimage.html). Via the `shading` parameter, we can apply an illumination to the grid or image, which can help to create visually appealing maps. By passing the argument **+a**, we use an azimuth of -45 degrees. Run the next cod ...
```
I am wondering if and how it is possible to add links to the other Jupyter notebooks or scripts (if this is possible, maybe we can also do this in a common PR).

jhtong33 replied:
I'm not sure I fully understand your point. Are you suggesting that this information could be linked to the current gallery?

yvonnefroehlich replied:
Hm. What do you mean with "current gallery"? 

We partly mention other tutorials, which is good. I feel it would be nice to directly insert the link. But we have the URL from GitHub (https://github.com/GenericMappingTools/agu24workshop/blob/main/book/tut0XYZ.ipynb) as well as the one from the  Jupyter book (https://www.generic-mapping-tools.org/agu24workshop/XYZ.html). So depending on which website people are, the link does lead them to another "base" website. As the Jupyter book website is the "official" website for this workshop and GitHub is more for development, maybe using the URL to the Jupyter book website makes more sense here?

jhtong33 replied:
Now I understand what you mean. 
> As the Jupyter book website is the "official" website for this workshop and GitHub is more for development, maybe using the URL to the Jupyter book website makes more sense here?

YES. Providing the URL to the Jupyter Book website will make it more convenient and readable.

jhtong33 committed:
Merge branch 'main' into add-jn-geo-seis

jhtong33 committed:
add tutorial 3 link

yvonnefroehlich committed:
Remove tut4

yvonnefroehlich committed:
Merge branch 'main' into add-jn-geo-seis

weiji14 commented on book/tut04_geophysics.ipynb:
```diff
@@ -0,0 +1,694 @@
+    "\n",
+    "1. `data` or `x`, `y`, `z`: Input text file. Importantly, if you use `data`, the columns order must be longitude, latitude, and data. Otherwise, you should use `x`, `y`, and `z` to individually define the dataset. \n",
+    "2. `region`: Specifies the grid's geographical region in the form `[lon_min, lon_max, lat_min, lat_max]`.\n",
+    "3. `spacing`: Grid spacing for the output grid. In our case, the spacing of original dataset is `5m`, we do not interpolate to better resolution, otherwise, it might cause data distortion. "
```
Best to clarify that this is 5 arc-minutes and not 5 metres - https://docs.generic-mapping-tools.org/6.5/xyz2grd.html#i

```suggestion
    "3. `spacing`: Grid spacing for the output grid. In our case, the spacing of original dataset is `5m` (5 arc-minutes), we do not interpolate to better resolution, otherwise, it might cause data distortion. "
```

weiji14 commented on book/tut04_geophysics.ipynb:
```diff
@@ -0,0 +1,694 @@
+    "## 2. Create a contour map: [`pygmt.Figure.grdcontour`](https://www.pygmt.org/v0.13.0/api/generated/pygmt.Figure.grdcontour.html)\n",
+    "To plot grids or images as contour lines you can use [`pygmt.Figure.grdcontour`](https://www.pygmt.org/v0.13.0/api/generated/pygmt.Figure.grdcontour.html). You need to define:\n",
+    "\n",
+    "1. `grid`: Access the remote datasets or providing your dataset as `xarray.DataArray`.\n",
```
```suggestion
    "1. `grid`: Access the remote datasets or provide your dataset as an `xarray.DataArray`.\n",
```

weiji14 commented on book/tut04_geophysics.ipynb:
```diff
@@ -0,0 +1,694 @@
+    "3. `annotation`: Annotate contour levels.\n",
+    "4. `limit`: Draw contours below low or above high, e.g., [-4000, 0] means drawing contour lines below sea level and above -4000 m.\n",
+    "\n",
+    "Before we plot the grid as contour lines, we plot the grid with color-coding using [`pygmt.Figure.grdimage`](https://www.pygmt.org/v0.13.0/api/generated/pygmt.Figure.grdimage.html). Via the `shading` parameter, we can apply an illumination to the grid or image, which can help to create visually appealing maps. By passing the argument **+a**, we use an azimuth of -45 degrees. Run the next cod ...
```
Can use relative link to the notebook - https://jupyterbook.org/en/stable/content/references.html#reference-content-files

```suggestion
    "Before we plot the grid as contour lines, we plot the grid with color-coding using [`pygmt.Figure.grdimage`](https://www.pygmt.org/v0.13.0/api/generated/pygmt.Figure.grdimage.html). Via the `shading` parameter, we can apply an illumination to the grid or image, which can help to create visually appealing maps. By passing the argument **+a**, we use an azimuth of -45 degrees. Run the next code cell twice, first without and second with illumination, to see the difference. For a more complicated illumination, please refer to [`pygmt.grdgradient`](https://www.pygmt.org/v0.13.0/api/generated/pygmt.grdgradient.html), which was already introduced in [Tutorial 3](../tut03_spe_xarray)."
```

weiji14 commented on book/tut04_geophysics.ipynb:
```diff
@@ -0,0 +1,694 @@
+   "id": "833bb3c5-70b7-4090-9418-26f19ae994a8",
+   "metadata": {},
+   "source": [
+    "Plotting the elevation along the profile and coloring land and ocean in lightgray or lightblue, respectively:"
```
Reverse the order, following the code cell below

```suggestion
    "Plotting the elevation along the profile and coloring ocean and land in lightblue and lightgray, respectively:"
```

weiji14 commented on book/tut04_geophysics.ipynb:
```diff
@@ -0,0 +1,694 @@
+   "id": "99d1d66d-b972-401a-8908-7a5783cae189",
+   "metadata": {},
+   "source": [
+    "Now you can plot the earthquakes of the two datasets on top of our map, e.g., using circles for the events with lower and stars for the events with higher magnitude. Additionally, you can add color- and size-coding to include epicentral distance and magnitude of the earthquakes. Take a look at what you have learned in Tutorial 2 🙂 and the PyGMT Tutorial [Plotting data points](https://www.pyg ...
```
```suggestion
    "Now you can plot the earthquakes of the two datasets on top of our map, e.g., using circles for the events with lower and stars for the events with higher magnitude. Additionally, you can add color- and size-coding to include epicentral distance and magnitude of the earthquakes. Take a look at what you have learned in [Tutorial 2](../tut02_spe_pd_gpd) 🙂 and the PyGMT Tutorial [Plotting data points](https://www.pygmt.org/v0.13.0/tutorials/basics/plot.html) as an orientation."
```

weiji14 reviewed (approved):
Wonderful, thanks @yvonnefroehlich and @jhtong33 for working on this! Sorry for leaving this last, Seismology isn't my area of expertise, so I was hoping someone would jump in to review, but I trust that this is ok. Personally, I would have preferred it if the .txt and .csv file were not in the repository since they are big files, but hopefully everyone has good internet connection during the workshop :laughing:

I've got some minor suggestions below, which either of you can apply, and then I will try to merge this by tomorrow. After this, we can work on polishing all of the notebooks to have a consistent header, title and such.

jhtong33 commented:
> Personally, I would have preferred it if the .txt and .csv file were not in the repository since they are big files, but hopefully everyone has good internet connection during the workshop 😆

If needed, I can adjust the period of the earthquake catalog, which might reduce the file size. (current size: 63 kb)

jhtong33 committed:
Apply suggestions from code review

Co-authored-by: Wei Ji <23487320+weiji14@users.noreply.github.com>

yvonnefroehlich commented:
> Wonderful, thanks @yvonnefroehlich and @jhtong33 for working on this! Sorry for leaving this last, Seismology isn't my area of expertise, so I was hoping someone would jump in to review, but I trust that this is ok.

Now worries @weiji14, @Esteban82 made already some helpful suggestions! And somethings it's actually also good to have a review from another field and also good to a have native-speaking person reading through the documentation.

> Personally, I would have preferred it if the .txt and .csv file were not in the repository since they are big files, but hopefully everyone has good internet connection during the workshop 😆

Yeah! Also was thinking about what is better: people download the data during the tutorials _or_ we provide data files within the repo, when creating the pands / geopandas tutorial. Downloading the data directly underlines that PyGMT integrates smoothly in the complete analysis workflow, and we keep the repo size small. Regarding a stable internet connection, I feel (hope) this should work (I have never been to AGU or any other big conference yet, but a working (acceptable fast) internet connection is essential for a nicely running conference) .

Regarding the elevation data: The "normal" way doing this would be to simply download the built-in GMT remote dataset for the Earth relief into a `xarray.DataArray` and using it for the further processing and analysis. However, in this tutorial, we like to show how people can convert any specific grid which is provided as xyz columns into a [Py]GMT-ready grid. Thus, the elevation data was written to a txt file with xyz columns, and in the tutorial we convert it back to a [Py]GMT-ready grid via `xyz2grd`. 
  
Regarding the seismicity data: ObsPy is required to request and download this data. If we include this part within the tutorial, ObsPy will become a dependency, which people need to install. This was probably the reason why Jing-Hui downloaded the data externally and wrote it to a CSV file for usage in the tutorial (reading via pandas).

> I've got some minor suggestions below, which either of you can apply, and then I will try to merge this by tomorrow. After this, we can work on polishing all of the notebooks to have a consistent header, title and such.

Sounds good!

jhtong33 commented:
> Regarding the elevation data: The "normal" way doing this would be to simply download the built-in GMT remote dataset for the Earth relief into a xarray.DataArray and using it for the further processing and analysis. However, in this tutorial, we like to show how people can convert any specific grid which is provided as xyz columns into a [Py]GMT-ready grid. Thus, the elevation data was written to a txt file with xyz columns, and in the tutorial we convert it back to a [Py]GMT-ready grid via xyz2grd.

Yvonne mentioned a critical point: many current datasets (e.g., crustal thickness and seismic tomography) are available only as xyz columns rather than pre-processed GMT grids. If beginners fail to correctly convert xyz data into a grid, using `grdimage ` will not produce accurate results. That’s why providing the xyz file is essential. 

> Regarding the seismicity data: ObsPy is required to request and download this data. If we include this part within the tutorial, ObsPy will become a dependency, which people need to install. This was probably the reason why Jin-Hui downloaded the data externally and wrote it to a CSV file for usage in the tutorial (reading via pandas).

If including Obspy, installing and downloading [IRIS dataset](https://docs.obspy.org/packages/autogen/obspy.clients.fdsn.client.Client.get_events.html) more relies on internet. 
I can mitigate this by reducing the file size, perhaps by limiting the earthquake data to a shorter time period.

jhtong33 committed:
EQ catalog smaller, and modify the searching criteria and period

weiji14 commented:
> > Regarding the elevation data: The "normal" way doing this would be to simply download the built-in GMT remote dataset for the Earth relief into a xarray.DataArray and using it for the further processing and analysis. However, in this tutorial, we like to show how people can convert any specific grid which is provided as xyz columns into a [Py]GMT-ready grid. Thus, the elevation data was written to a txt file with xyz columns, and in the tutorial we convert it back to a [Py]GMT-ready grid via xyz2grd.
> 
> Yvonne mentioned a critical point: many current datasets (e.g., crustal thickness and seismic tomography) are available only as xyz columns rather than pre-processed GMT grids. If beginners fail to correctly convert xyz data into a grid, using `grdimage ` will not produce accurate results. That’s why providing the xyz file is essential.

Ah I see, then teaching xyz2grd would make sense here. Thanks for the extra context, this is why having people working in the same field makes sense :wink:

> > Regarding the seismicity data: ObsPy is required to request and download this data. If we include this part within the tutorial, ObsPy will become a dependency, which people need to install. This was probably the reason why Jin-Hui downloaded the data externally and wrote it to a CSV file for usage in the tutorial (reading via pandas).
> 
> If including Obspy, installing and downloading [IRIS dataset](https://docs.obspy.org/packages/autogen/obspy.clients.fdsn.client.Client.get_events.html) more relies on internet. I can mitigate this by reducing the file size, perhaps by limiting the earthquake data to a shorter time period.

The earthquake data in CSV format is actually not too big (64KB before, now 29KB). It's more the south_america_topography_05m.txt file that's a little large (5.0MB) :slightly_smiling_face:, but I think it's ok to just keep it in the repository for now instead of changing the workflow. Also ok with having ObsPy included in the dependency list, but that can be done in a follow-up PR (only if there's time, otherwise the PR is ok as is).

@jhtong33, I'll leave this up for a few more hours in case there's any small changes you still want to make, and will merge once you hit approve (or by the end of my work day in ~7 hours).

weiji14 commented:
Tutorial is now up at https://www.generic-mapping-tools.org/agu24workshop/tut04_geophysics.html :tada:
